### PR TITLE
API: Validate redirect_to cookie has valid (Grafana) url 

### LINKED ISSUE #20641: Validate redirect_to cookie has valid (Grafana) url
When using the redirect_to cookie we should validate that it contains a valid http url: IE that it starts with Grafana app url.

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
This fix adds validation for the `redirect_to` cookie value and ignores invalid ones.

**Which issue(s) this PR fixes**:

Fixes #20641

**Special notes for your reviewer**:
Contrary to the indication in the issue description (proposing to check that the value starts with the Grafana app url), this fix allows only relative paths starting with Grafana app sub url (if that exists). The reason is that only the relative path is [set](https://github.com/grafana/grafana/blob/master/pkg/middleware/auth.go#L50) to the cookie. However, maybe there is a use case, that I'm not aware of, which requires an absolute url in the `redirect_to`.

### CODE REVIEW

File: pkg/api/login_test.go [hunk 1]
```diff
@@ -104,6 +123,197 @@ func TestLoginErrorCookieApiEndpoint(t *testing.T) {
...
+			sc.fakeReqNoAssertionsWithCookie("GET", sc.url, cookie).exec()
+			assert.Equal(t, c.status, sc.resp.Code)
+			if c.status == 302 {
+				location, ok := sc.resp.Header()["Location"]
+				assert.True(t, ok)
+				assert.Equal(t, location[0], c.url)
+			}
+
+			responseString, err := getBody(sc.resp)
+			assert.Nil(t, err)
```
Comment (aknuds1) [inline pkg/api/login_test.go]: Niggle, but could you use `assert.NoError` here? It's clearer.

File: pkg/api/login_test.go [hunk 2]
```diff
@@ -104,6 +123,197 @@ func TestLoginErrorCookieApiEndpoint(t *testing.T) {
...
+				Path:     setting.AppSubUrl + "/",
+				Secure:   hs.Cfg.CookieSecure,
+				SameSite: hs.Cfg.CookieSameSite,
+			}
+			sc.m.Post(sc.url, sc.defaultHandler)
+			sc.fakeReqNoAssertionsWithCookie("POST", sc.url, cookie).exec()
+			assert.Equal(t, sc.resp.Code, 200)
+
+			respJSON, err := simplejson.NewJson(sc.resp.Body.Bytes())
+			assert.Nil(t, err)
```
Comment (aknuds1) [inline pkg/api/login_test.go]: Same as above, I think assert.NoError would be clearer.

**Review by aknuds1 — APPROVED** [review]
LGTM, just a couple of nits.

**Review by marefr — APPROVED** [review]
Works as expected. But agree with Arve's suggestions.

@papagian  Are you planning to finish up #19744 as well when this have been merged?

**Comment by papagian** [comment]
> @papagian Are you planning to finish up #19744 as well when this have been merged?

There is an [open](https://github.com/grafana/grafana/pull/19787) PR for this, I can contribute to this one by adding the suggested changes.

**Comment by marefr** [comment]
@papagian yeah sounds good :+1:

**Comment by marefr** [comment]
@papagian can you please add labels and milestone to pr and issue

**Pull request merged**